In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)


import os
for dirname, _, filenames in os.walk('/kaggle/input'):0
    for filename in filenames:
        print(os.path.join(dirname, filename))


IndentationError: unexpected indent (3356492338.py, line 7)

In [3]:
!pip install mlxtend

   ---------------------------------------- 0.0/1.4 MB ? eta -:--:--
   ---------------------------------------- 1.4/1.4 MB 11.7 MB/s eta 0:00:00


In [4]:
import csv
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules

In [5]:
data = []
with open('Market_Basket_Optimisation.csv') as file:
    reader = csv.reader(file, delimiter=',')
    for row in reader:
        data +=[row]

In [6]:
data[1:10] #list of list

[['burgers', 'meatballs', 'eggs'],
 ['chutney'],
 ['turkey', 'avocado'],
 ['mineral water', 'milk', 'energy bar', 'whole wheat rice', 'green tea'],
 ['low fat yogurt'],
 ['whole wheat pasta', 'french fries'],
 ['soup', 'light cream', 'shallot'],
 ['frozen vegetables', 'spaghetti', 'green tea'],
 ['french fries']]

In [11]:
len(data)

7501

In [13]:
te = TransactionEncoder()
x = te.fit_transform(data)

In [15]:
x

array([[False,  True,  True, ...,  True, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       ...,
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False,  True, False]])

In [17]:
te.columns_

[' asparagus',
 'almonds',
 'antioxydant juice',
 'asparagus',
 'avocado',
 'babies food',
 'bacon',
 'barbecue sauce',
 'black tea',
 'blueberries',
 'body spray',
 'bramble',
 'brownies',
 'bug spray',
 'burger sauce',
 'burgers',
 'butter',
 'cake',
 'candy bars',
 'carrots',
 'cauliflower',
 'cereals',
 'champagne',
 'chicken',
 'chili',
 'chocolate',
 'chocolate bread',
 'chutney',
 'cider',
 'clothes accessories',
 'cookies',
 'cooking oil',
 'corn',
 'cottage cheese',
 'cream',
 'dessert wine',
 'eggplant',
 'eggs',
 'energy bar',
 'energy drink',
 'escalope',
 'extra dark chocolate',
 'flax seed',
 'french fries',
 'french wine',
 'fresh bread',
 'fresh tuna',
 'fromage blanc',
 'frozen smoothie',
 'frozen vegetables',
 'gluten free bar',
 'grated cheese',
 'green beans',
 'green grapes',
 'green tea',
 'ground beef',
 'gums',
 'ham',
 'hand protein bar',
 'herb & pepper',
 'honey',
 'hot dogs',
 'ketchup',
 'light cream',
 'light mayo',
 'low fat yogurt',
 'magazines',
 'mashe

In [21]:
df = pd.DataFrame(x, columns=te.columns_)

C:\Users\fahee\anaconda3\Lib\site-packages\executing\executing.py:713: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
  right=ast.Str(s=sentinel),
C:\Users\fahee\anaconda3\Lib\ast.py:587: DeprecationWarning: Attribute s is deprecated and will be removed in Python 3.14; use value instead
  return Constant(*args, **kwargs)
C:\Users\fahee\anaconda3\Lib\site-packages\executing\executing.py:713: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
  right=ast.Str(s=sentinel),
C:\Users\fahee\anaconda3\Lib\ast.py:587: DeprecationWarning: Attribute s is deprecated and will be removed in Python 3.14; use value instead
  return Constant(*args, **kwargs)


NameError: name 'pd' is not defined

In [23]:
df

C:\Users\fahee\anaconda3\Lib\site-packages\executing\executing.py:713: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
  right=ast.Str(s=sentinel),
C:\Users\fahee\anaconda3\Lib\ast.py:587: DeprecationWarning: Attribute s is deprecated and will be removed in Python 3.14; use value instead
  return Constant(*args, **kwargs)
C:\Users\fahee\anaconda3\Lib\site-packages\executing\executing.py:713: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
  right=ast.Str(s=sentinel),
C:\Users\fahee\anaconda3\Lib\ast.py:587: DeprecationWarning: Attribute s is deprecated and will be removed in Python 3.14; use value instead
  return Constant(*args, **kwargs)


NameError: name 'df' is not defined

In [25]:
# Finding frequent itemsets
freq_itemset = apriori(df, min_support=0.01, use_colnames=True)

C:\Users\fahee\anaconda3\Lib\site-packages\executing\executing.py:713: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
  right=ast.Str(s=sentinel),
C:\Users\fahee\anaconda3\Lib\ast.py:587: DeprecationWarning: Attribute s is deprecated and will be removed in Python 3.14; use value instead
  return Constant(*args, **kwargs)
C:\Users\fahee\anaconda3\Lib\site-packages\executing\executing.py:713: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
  right=ast.Str(s=sentinel),
C:\Users\fahee\anaconda3\Lib\ast.py:587: DeprecationWarning: Attribute s is deprecated and will be removed in Python 3.14; use value instead
  return Constant(*args, **kwargs)


NameError: name 'df' is not defined

0.01 means 1% of data size(7501) is 75. so single item should come atleast 75 times then take those items in freq_itemset

In [67]:
freq_itemset

support                                 itemsets
0    0.020397                                (almonds)
1    0.033329                                (avocado)
2    0.010799                         (barbecue sauce)
3    0.014265                              (black tea)
4    0.011465                             (body spray)
..        ...                                      ...
252  0.011065       (milk, ground beef, mineral water)
253  0.017064  (ground beef, spaghetti, mineral water)
254  0.015731         (milk, spaghetti, mineral water)
255  0.010265    (spaghetti, mineral water, olive oil)
256  0.011465     (spaghetti, mineral water, pancakes)

[257 rows x 2 columns]

In [68]:
from mlxtend.frequent_patterns import association_rules


In [69]:
rules = association_rules(freq_itemset, metric="confidence", min_threshold=0.10, num_itemsets=10)


In [70]:
rules = rules[['antecedents','consequents','support','confidence']]
rules

antecedents                 consequents   support  \
0                     (avocado)             (mineral water)  0.011598   
1                     (burgers)                      (cake)  0.011465   
2                        (cake)                   (burgers)  0.011465   
3                     (burgers)                 (chocolate)  0.017064   
4                   (chocolate)                   (burgers)  0.017064   
..                          ...                         ...       ...   
315                 (olive oil)  (mineral water, spaghetti)  0.010265   
316  (mineral water, spaghetti)                  (pancakes)  0.011465   
317       (spaghetti, pancakes)             (mineral water)  0.011465   
318   (mineral water, pancakes)                 (spaghetti)  0.011465   
319                  (pancakes)  (mineral water, spaghetti)  0.011465   

     confidence  
0      0.348000  
1      0.131498  
2      0.141447  
3      0.195719  
4      0.104150  
..          ...  
315    0.155870  
316    0.191964  
317    0.455026  
318    0.339921  
319    0.120617  

[320 rows x 4 columns]

In [71]:
rules[rules['antecedents'] == {'cake'}]['consequents']

2               (burgers)
25            (chocolate)
27                 (eggs)
29         (french fries)
31    (frozen vegetables)
33            (green tea)
35                 (milk)
37        (mineral water)
39             (pancakes)
41            (spaghetti)
Name: consequents, dtype: object